core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/math/median.kt

```diff
@@ -1,10 +1,169 @@
 package org.jetbrains.kotlinx.dataframe.math
 
+import io.github.oshai.kotlinlogging.KotlinLogging
+import org.jetbrains.kotlinx.dataframe.api.isNaN
+import org.jetbrains.kotlinx.dataframe.impl.aggregation.aggregators.CalculateReturnType
+import org.jetbrains.kotlinx.dataframe.impl.canBeNaN
+import org.jetbrains.kotlinx.dataframe.impl.isIntraComparable
+import org.jetbrains.kotlinx.dataframe.impl.isPrimitiveNumber
+import org.jetbrains.kotlinx.dataframe.impl.nothingType
+import org.jetbrains.kotlinx.dataframe.impl.renderType
+import org.jetbrains.kotlinx.dataframe.math.quickSelect
+import java.math.BigDecimal
+import java.math.BigInteger
 import kotlin.reflect.KType
+import kotlin.reflect.full.withNullability
 import kotlin.reflect.typeOf
 
+private val logger = KotlinLogging.logger { }
+
 // TODO median always returns the same type, but this can be confusing for iterables of even length
 // TODO (e.g. median of [1, 2] should be 1.5, but the type is Int, so it returns 1), Issue #558
+
+/**
+ * Returns the median of the comparable input:
+ * - `null` if empty
+ * - `Double` if primitive number
+ * - `Double.NaN` if ![skipNaN] and contains NaN
+ * - (lower) middle else
+ *
+ * TODO migrate back to percentile when it's flexible enough
+ */
 @PublishedApi
-internal inline fun <reified T : Comparable<T>> Iterable<T?>.median(type: KType = typeOf<T>()): T? =
-    percentile(50.0, type)
+internal fun <T : Comparable<T>> Sequence<T>.medianOrNull(type: KType, skipNaN: Boolean): Any? {
+    when {
+        type.isMarkedNullable ->
+            error("Encountered nullable type ${renderType(type)} in median function. This should not occur.")
+
+        !type.isIntraComparable() ->
+            error(
+                "Unable to compute the median for ${
+                    renderType(type)
+                }. Only primitive numbers or self-comparables are supported.",
+            )
+
+        type == typeOf<BigDecimal>() || type == typeOf<BigInteger>() ->
+            throw IllegalArgumentException(
+                "Cannot calculate the median for big numbers in DataFrame. Only primitive numbers are supported.",
+            )
+
+        type == typeOf<Long>() ->
+            logger.warn { "Converting Longs to Doubles to calculate the median, loss of precision may occur." }
```

Comment:
probably debug mode?

Reply:
It's just a warning logged right? users can disable it by changing the log level for `org.jetbrains.kotlinx.dataframe.math.MedianKt` on their end or by converting their columns to `Double` manually first. It's not a heavy check, so I think debug mode is not needed.